media/js/base/stub-attribution.js

```diff
@@ -0,0 +1,264 @@
+
+                version = $(this).data('downloadVersion');
+                // Currently only Windows 32bit uses the stub installer, but this could
+                // easily change in the future so we'll make that bet now.
+                if (version && (version === 'win' || version === 'win64')) {
+                    this.href = Mozilla.StubAttribution.appendToDownloadURL(this.href, data);
+                }
+            }
+        });
+    };
+
+    /**
+     * Appends stub attribution data as URL parameters.
+     * Note: data is already URI encoded when returned via the service.
+     * @param {String url - URL to append data to.
+     * @param {Object} data - attribution_code, attribution_sig.
+     * @return {String} url + additional parameters.
+     */
+    StubAttribution.appendToDownloadURL = function(url, data) {
+
+        if (!data.attribution_code || !data.attribution_sig) {
+            return url;
+        }
+
+        // append stub attribution query params.
+        $.each(data, function(key, val) {
+            if (key === 'attribution_code' || key === 'attribution_sig') {
+                url += (url.indexOf('?') > -1 ? '&' : '?') + key + '=' + val;
+            }
+        });
+
+        return url;
+    };
+
+    /**
+     * Handles XHR request from `stub_attribution_code` service.
+     * @param {Object} data - attribution_code, attribution_sig.
+     */
+    StubAttribution.onRequestSuccess = function(data) {
+        if (data.attribution_code && data.attribution_sig) {
+            // Update download links on the current page.
+            StubAttribution.updateBouncerLinks(data);
+            // Store attribution data in a cookie should the user navigate.
+            StubAttribution.setCookie(data);
+        }
+    };
+
+    /**
+     * AJAX request to bedrock service to authenticate stub attribution request.
+     * @param {Object} data - utm params and referrer.
+     */
+    StubAttribution.requestAuthentication = function(data) {
+        var SERVICE_URL = window.location.protocol + '//' + window.location.host + '/en-US/firefox/stub_attribution_code/';
+        $.get(SERVICE_URL, data).done(StubAttribution.onRequestSuccess);
+    };
+
+    /**
+     * Gets utm parameters and referrer information from the web page if they exist.
+     * @param {String} ref - Optional referrer to facilitate testing.
+     * @return {Object} - Stub attribution data object.
+     */
+    StubAttribution.getAttributionData = function(ref) {
+        var params = new window._SearchParams().utmParams();
+        var referrer = typeof ref !== 'undefined' ? ref : document.referrer;
+        var utmCount = 0;
+
+        for (var utm in params) {
+            if (params.hasOwnProperty(utm)) {
+                utmCount += 1;
+            }
+        }
+
+        // if there are no utm params and no referrer, do nothing.
+        if (utmCount === 0 && (typeof referrer === 'undefined' || referrer === '')) {
```

Comment:
Apologies if this has been covered elsewhere, or is totally out of scope, but do we care about links coming from marketing emails or other non-web page sources? I'm not sure those have a referrer...